tools/approval-service/internal/service/ttleventcache.go

```diff
@@ -0,0 +1,152 @@
+	closed          bool
+}
+
+// Option configures TTLEventCache construction
+type Option func(*TTLEventCache) error
+
+// WithCleanupInterval allows overriding the tick interval (used by janitor for cleanup) via option.
+// Validates duration > 0.
+func WithCleanupInterval(interval time.Duration) Option {
+	return func(c *TTLEventCache) error {
+		if interval <= 0 {
+			return fmt.Errorf("invalid interval")
+		}
+		c.cleanupInterval = interval
+		return nil
+	}
+}
+
+// MakeTTLEventCache creates a new TTLEventCache.
+// ttl: how long to prevent re-processing after an arrival or after finish().
+// Use options for additional configuration (e.g., WithCleanupInterval).
+func MakeTTLEventCache(ttl time.Duration, opts ...Option) (*TTLEventCache, func() error, error) {
+	// set a logical default
+	if ttl <= 0 {
+		ttl = 15 * time.Second
+	}
+
+	// default tick: ttl/4 for moderate TTLs, minimum 1s
+	defaultCleanupInterval := func(ttl time.Duration) time.Duration {
+		if ttl > 4*time.Second {
+			return ttl / 4
+		}
+		return 1 * time.Second
+	}
+
+	c := &TTLEventCache{
+		items:           make(map[string]time.Time),
+		ttl:             ttl,
+		cleanupInterval: defaultCleanupInterval(ttl),
+		stop:            make(chan struct{}),
+		done:            make(chan struct{}),
+	}
+
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	cleanup := func() error {
+		_ = c.Stop(context.Background())
+		return nil
+	}
+
+	// start evictor
+	go c.evictor()
+	return c, cleanup, nil
+}
+
+// TryAdd implements debounce-on-arrival semantics.
+// Returns true if the caller should process the event; false if unexpired entry exists.
+func (c *TTLEventCache) TryAdd(eventID string) bool {
+	now := time.Now()
+
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if c.closed {
+		return false
+	}
+
+	if expiry, ok := c.items[eventID]; ok {
+		// unexpired -> reject
+		if now.Before(expiry) {
+			return false
+		}
+
+		// expired -> delete and allow insertion
+		delete(c.items, eventID)
+	}
+
+	c.items[eventID] = now.Add(c.ttl)
+	return true
+}
+
+// Stop stops the cleaner (key evictor) and prevents further additions. It blocks until the cleaner exits or ctx is done.
+func (c *TTLEventCache) Stop(ctx context.Context) error {
+	c.mu.Lock()
+	if c.closed {
+		c.mu.Unlock()
+		return nil
+	}
+	c.closed = true
+	// signal cleaner
+	close(c.stop)
+	c.mu.Unlock()
+
+	select {
+	case <-c.done:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Len returns the number of entries currently in the cache. Useful for tests/metrics.
+func (c *TTLEventCache) Len() int {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	return len(c.items)
+}
+
+// evictor runs in a background goroutine and periodically removes expired entries from the cache.
+func (c *TTLEventCache) evictor() {
+	ticker := time.NewTicker(c.cleanupInterval)
+	defer ticker.Stop()
+	defer close(c.done)
+
+	for {
+		select {
+		case <-ticker.C:
+			now := time.Now()
```

Comment:
The ticker sends the current time each time it fires, so you can do `case now := <-ticker.C` and remove this line